SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/OwnershipLockingPolicy.cpp

```diff
@@ -0,0 +1,297 @@
+	}
+
+	if (MigrationLockElement* ActorLockingState = ActorToLockingState.Find(Actor))
+	{
+		++ActorLockingState->LockCount;
+	}
+	else
+	{
+		// We want to avoid memory leak if a locked actor is deleted.
+		// To do this, we register with the Actor OnDestroyed delegate with a function that cleans up the internal map.
+		if (!Actor->OnDestroyed.IsAlreadyBound(this, &UOwnershipLockingPolicy::OnExplicitlyLockedActorDeleted))
+		{
+			Actor->OnDestroyed.AddDynamic(this, &UOwnershipLockingPolicy::OnExplicitlyLockedActorDeleted);
+		}
+
+		AActor* OwnershipHierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+		AddOwnershipHierarchyRootInformation(OwnershipHierarchyRoot, Actor);
+		
+		ActorToLockingState.Add(Actor, MigrationLockElement{ 1, OwnershipHierarchyRoot });
+	}
+
+	UE_LOG(LogOwnershipLockingPolicy, Log, TEXT("Acquiring migration lock. "
+		"Actor: %s. Lock name: %s. Token %d: Locks held: %d."), *GetNameSafe(Actor), *DebugString, NextToken, ActorToLockingState.Find(Actor)->LockCount);
+	TokenToNameAndActor.Emplace(NextToken, LockNameAndActor{ MoveTemp(DebugString), Actor });
+	return NextToken++;
+}
+
+bool UOwnershipLockingPolicy::ReleaseLock(const ActorLockToken Token)
+{
+	const LockNameAndActor* NameAndActor = TokenToNameAndActor.Find(Token);
+	if (NameAndActor == nullptr)
+	{
+		UE_LOG(LogOwnershipLockingPolicy, Error, TEXT("Called ReleaseLock for unidentified Actor lock token. Token: %d."), Token);
+		return false;
+	}
+
+	AActor* Actor = NameAndActor->Actor;
+	const FString& Name = NameAndActor->LockName;
+	UE_LOG(LogOwnershipLockingPolicy, Log, TEXT("Releasing Actor migration lock. Actor: %s. Token: %d. Lock name: %s"), *Actor->GetName(), Token, *Name);
+
+	check(ActorToLockingState.Contains(Actor));
+
+	{
+		// Reduce the reference count and erase the entry if reduced to 0.
+		auto CountIt = ActorToLockingState.CreateKeyIterator(Actor);
+		MigrationLockElement& ActorLockingState = CountIt.Value();
+		if (ActorLockingState.LockCount == 1)
+		{
+			UE_LOG(LogOwnershipLockingPolicy, Log, TEXT("Actor migration no longer locked. Actor: %s"), *Actor->GetName());
+			Actor->OnDestroyed.RemoveDynamic(this, &UOwnershipLockingPolicy::OnExplicitlyLockedActorDeleted);
+			RemoveOwnershipHierarchyRootInformation(ActorLockingState.HierarchyRoot, Actor);
+			CountIt.RemoveCurrent();
+		}
+		else
+		{
+			--ActorLockingState.LockCount;
+		}
+	}
+
+	TokenToNameAndActor.Remove(Token);
+
+	return true;
+}
+
+bool UOwnershipLockingPolicy::IsLocked(const AActor* Actor) const
+{
+	if (Actor == nullptr)
+	{
+		UE_LOG(LogOwnershipLockingPolicy, Warning, TEXT("IsLocked called for nullptr"));
+		return false;
+	}
+
+	// Is this Actor explicitly locked or on a locked hierarchy ownership path.
+	if (IsExplicitlyLocked(Actor) || IsLockedHierarchyRoot(Actor))
+	{
+		return true;
+	}
+
+	// Is the hierarchy root of this Actor explicitly locked or on a locked hierarchy ownership path.
+	if (AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor))
```

Comment:
Could just return `IsLocked( SpatialGDK::GetHierarchyRoot(Actor) )` here.

Reply:
Oh wait, that'll error when the root eventually resolves to null

Reply:
If you're a root Actor you'll get a warning from passing a nullptr to IsLocked